### PR TITLE
Market Data Tab

### PR DESCRIPTION
**Pull request by muliswilliam**
This PR adds a new tab in Wallet UI: Market. The tab shows coin markets data from coingecko.
The following data fields are displayed:
- Asset Icon
- Asset Name
- Asset Symbol
- Price
- 24 hr percentage
- Market cap
- Volume

The UI also provides filter and search functionality. The available filters are:
- All Assets - displays top 250 assets from Coingecko
- Tradable - displays assets that support swap in brave wallet

MIssing from current version:
- Add to wishlist(Will be added when a way to store wishlist is created in the backend)
- A line graph showing price changes in a specified period e.g., 24hr(Will be added when data is available from backend service)

**Note: This PR will only be merged after this PR https://github.com/brave/brave-core/pull/12510 is merged since this feature has a dependency on it.**

Figma design: https://www.figma.com/file/jZrVG7dsi5npXFTDC6LDGn/Crypto?node-id=95%3A9890

Resolves https://github.com/brave/brave-browser/issues/20674

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/784) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

https://user-images.githubusercontent.com/48117473/158770219-0d394868-3bc4-45b0-b975-69d75001e0ca.mov

### CODE REVIEW

**Comment by diracdeltas** [comment]
@muliswilliam this PR has a ton of unrelated changes - you probably need to rebase it.

also i'm not sure what this feature is, but if it involves adding new dependencies or network requests, you should open a security review. https://github.com/brave/brave-browser/wiki/Security-reviews

**Comment by muliswilliam** [comment]
@diracdeltas Rebased and created a security review request here: https://github.com/brave/security/issues/784

File: components/brave_wallet/browser/asset_ratio_response_parser.cc
Comment (diracdeltas) [inline components/brave_wallet/browser/asset_ratio_response_parser.cc:425]: i think this should also wait for @spylogsster 's work on rust json validation before being merged. cc @bridiver

File: package.json
Comment (diracdeltas) [inline package.json:348]: why were these versions lowered?

**Comment by diracdeltas** [comment]
Could we load this entire feature inside a `chrome-untrusted://wallet-market-data` frame instead of at the top level? Since it adds and requires a bunch of dependencies, this would help sandbox the wallet from security issues in the dependencies.

From looking at the Figma spec, this seems easy to do without having any impact on the UX.

https://chromium.googlesource.com/chromium/src/+/master/docs/chrome_untrusted.md#Embed-chrome_untrusted_in-chrome_WebUIs

**Comment by diracdeltas** [comment]
per https://github.com/brave/security/issues/663#issuecomment-967770990, i think we need a "powered by coingecko" notice somewhere on this if we're using their API? not sure if that's already in the designs.

File: components/brave_wallet/browser/brave_wallet_constants.cc
Comment (muliswilliam) [reply components/brave_wallet/browser/brave_wallet_constants.cc:10]: Yes, this will be reverted before merging. This is temporary to allow for testing

File: package.json
Comment (muliswilliam) [reply package.json:348]: I will revert this to match those on `master`.

**Comment by muliswilliam** [comment]
> Could we load this entire feature inside a `chrome-untrusted://wallet-market-data` frame instead of at the top level? Since it adds and requires a bunch of dependencies, this would help sandbox the wallet from security issues in the dependencies.
> 
> From looking at the Figma spec, this seems easy to do without having any impact on the UX.
> 
> https://chromium.googlesource.com/chromium/src/+/master/docs/chrome_untrusted.md#Embed-chrome_untrusted_in-chrome_WebUIs

Is this a frontend change? I do not know how it works, I will need guidance on the same.

**Comment by muliswilliam** [comment]
> per [brave/security#663 (comment)](https://github.com/brave/security/issues/663#issuecomment-967770990), i think we need a "powered by coingecko" notice somewhere on this if we're using their API? not sure if that's already in the designs.

Good idea,  @jamesmudgett can this be added in the design?

**Comment by diracdeltas** [comment]
> Is this a frontend change? I do not know how it works, I will need guidance on the same.

it does require some cpp changes i believe, perhaps @petemill could help here?

**Review by diracdeltas — CHANGES_REQUESTED** [review]
see comments above

**Comment by muliswilliam** [comment]
> per [brave/security#663 (comment)](https://github.com/brave/security/issues/663#issuecomment-967770990), i think we need a "powered by coingecko" notice somewhere on this if we're using their API? not sure if that's already in the designs.
@diracdeltas I added "powered by coingecko"
Screenshot 2022-03-30 at 21 48 12

**Comment by nvonpentz** [comment]
Rebased to remove the backend code that's included in https://github.com/brave/brave-core/pull/12510, and changed the base branch from `master` to `markets-ratios-proxy` which is the branch with the backend logic.

We can merge this branch into `markets-ratios-proxy`, centralize the reviews and testing on that branch, then merge that into `master`.

**Comment by nvonpentz** [comment]
Merging this to the backend code. Some or all of the CI failures should be resolved after that.  We can iterate on both frontend and backend changes in that one PR.

**Pull request merged**
